File: crates/nitpick/src/scenarios/counter.rs
//! Simple counter scenario: contended increment/decrement of a document value.
//!
//! Two documents: `linear` (incremented each transaction) and `triangle`
//! (triangle += linear each transaction).
//!
//! Invariant: `triangle == (linear + 1) * linear / 2`
//!
//! This implies transactions are serialized -- no two transactions can read
//! the same starting snapshot and both commit successfully.

use std::sync::LazyLock;

use anyhow::bail;
use application::Application;
use async_trait::async_trait;
use common::{
    assert_obj,
    types::TableName,
    value::{
        ConvexObject,
        ConvexValue,
        ResolvedDocumentId,
        TableNamespace,
    },
};
use database::{
    TableModel,
    TestFacingModel,
    UserFacingModel,
};
use futures::{
    future::BoxFuture,
    FutureExt,
};
use keybroker::Identity;
use runtime::testing::TestRuntime;

use crate::framework::scenario::{
    Scenario,
    TestRun,
};

static TABLE: LazyLock<TableName> = LazyLock::new(|| "counter".parse().unwrap());

#[derive(Clone)]
pub struct CounterScenario;

#[async_trait]
impl Scenario for CounterScenario {
    type TestRun = CounterRun;

    fn name(&self) -> &'static str {
        "counter"
    }

    async fn start_run(
        &self,
        _rt: TestRuntime,
        application: &Application<TestRuntime>,
    ) -> anyhow::Result<CounterRun> {
        let mut tx = application.begin(Identity::system()).await?;
        TableModel::new(&mut tx)
            .insert_table_metadata(TableNamespace::test_user(), &TABLE)
            .await?;
        let linear_id = TestFacingModel::new(&mut tx)
            .insert(&TABLE, object_from_count(0))
            .await?;
        let triangle_id = TestFacingModel::new(&mut tx)
            .insert(&TABLE, object_from_count(0))
            .await?;
        application.commit_test(tx).await?;
        tracing::info!(
            "[counter] Initialized: linear={linear_id}, triangle={triangle_id}"
        );
        Ok(CounterRun {
            linear_id,
            triangle_id,
        })
    }
}

pub struct CounterRun {
    linear_id: ResolvedDocumentId,
    triangle_id: ResolvedDocumentId,
}

#[async_trait]
impl TestRun for CounterRun {
    type Output = i64;

    fn run_transaction(
        &self,
        _rt: TestRuntime,
        application: &Application<TestRuntime>,
    ) -> BoxFuture<'static, anyhow::Result<()>> {
        let linear_id = self.linear_id;
        let triangle_id = self.triangle_id;
        let application = application.clone();
        let future = async move {
            let mut tx = application.begin(Identity::system()).await?;

            let linear_doc = tx
                .get(linear_id)
                .await?
                .ok_or_else(|| anyhow::anyhow!("Linear document not found"))?;
            let linear = count_from_object(linear_doc.value())?;

            let triangle_doc = tx
                .get(triangle_id)
                .await?
                .ok_or_else(|| anyhow::anyhow!("Triangle document not found"))?;
            let triangle = count_from_object(triangle_doc.value())?;

            UserFacingModel::new_root_for_test(&mut tx)
                .replace(linear_id.into(), object_from_count(linear + 1))
                .await?;

            let linear_doc = tx
                .get(linear_id)
                .await?
                .ok_or_else(|| anyhow::anyhow!("Linear document not found after replace"))?;
            let new_linear = count_from_object(linear_doc.value())?;
            anyhow::ensure!(
                new_linear == linear + 1,
                "Within-transaction read inconsistency"
            );

            UserFacingModel::new_root_for_test(&mut tx)
                .patch(
                    triangle_id.into(),
                    object_from_count(triangle + new_linear).into(),
                )
                .await?;

            application.commit_test(tx).await?;
            Ok(())
        };
        future.boxed()
    }

    async fn validate(
        &self,
        application: &Application<TestRuntime>,
    ) -> anyhow::Result<()> {
        let mut tx = application.begin(Identity::system()).await?;
        let linear_doc = tx
            .get(self.linear_id)
            .await?
            .ok_or_else(|| anyhow::anyhow!("Linear document not found"))?;
        let triangle_doc = tx
            .get(self.triangle_id)
            .await?
            .ok_or_else(|| anyhow::anyhow!("Triangle document not found"))?;
        let linear = count_from_object(linear_doc.value())?;
        let triangle = count_from_object(triangle_doc.value())?;
        let expected_triangle = (linear + 1) * linear / 2;
        anyhow::ensure!(
            expected_triangle == triangle,
            "Invariant violated: linear={linear}, triangle={triangle}, expected={expected_triangle}"
        );
        Ok(())
    }

    async fn finalize(
        &self,
        application: &Application<TestRuntime>,
    ) -> anyhow::Result<i64> {
        let mut tx = application.begin(Identity::system()).await?;
        let triangle_doc = tx
            .get(self.triangle_id)
            .await?
            .ok_or_else(|| anyhow::anyhow!("Triangle document not found"))?;
        let triangle = count_from_object(triangle_doc.value())?;
        tracing::info!("[counter] Final triangle value: {triangle}");
        Ok(triangle)
    }
}

fn object_from_count(i: i64) -> ConvexObject {
    assert_obj!("count" => i)
}

fn count_from_object(o: &ConvexObject) -> anyhow::Result<i64> {
    let count = o
        .get("count")
        .ok_or_else(|| anyhow::anyhow!("count field not found"))?;
    if let ConvexValue::Int64(m) = count {
        if *m < 0 {
            bail!("Negative counter: {m}")
        }
        Ok(*m)
    } else {
        bail!("count field is not Int64: {count:?}")
    }
}

File: npm-packages/simulation/convex/counter.ts
import { v } from "convex/values";
import { mutation, query } from "./_generated/server";

export const setup = mutation({
  args: {},
  returns: v.object({
    linearId: v.id("counterLinear"),
    triangleId: v.id("counterTriangle"),
  }),
  handler: async (ctx) => {
    const linearId = await ctx.db.insert("counterLinear", { count: 0 });
    const triangleId = await ctx.db.insert("counterTriangle", { count: 0 });
    return { linearId, triangleId };
  },
});

export const increment = mutation({
  args: {
    linearId: v.id("counterLinear"),
    triangleId: v.id("counterTriangle"),
  },
  returns: v.null(),
  handler: async (ctx, args) => {
    const linearDoc = await ctx.db.get(args.linearId);
    if (!linearDoc) throw new Error("Linear doc not found");
    const triangleDoc = await ctx.db.get(args.triangleId);
    if (!triangleDoc) throw new Error("Triangle doc not found");

    const newLinear = linearDoc.count + 1;
    await ctx.db.replace(args.linearId, { count: newLinear });

    // Verify within-transaction read consistency.
    const check = await ctx.db.get(args.linearId);
    if (!check || check.count !== newLinear) {
      throw new Error("Within-transaction read inconsistency!");
    }

    await ctx.db.replace(args.triangleId, {
      count: triangleDoc.count + newLinear,
    });
    return null;
  },
});

export const validate = query({
  args: {
    linearId: v.id("counterLinear"),
    triangleId: v.id("counterTriangle"),
  },
  returns: v.object({ linear: v.number(), triangle: v.number() }),
  handler: async (ctx, args) => {
    const linearDoc = await ctx.db.get(args.linearId);
    if (!linearDoc) throw new Error("Linear doc not found");
    const triangleDoc = await ctx.db.get(args.triangleId);
    if (!triangleDoc) throw new Error("Triangle doc not found");

    const linear = linearDoc.count;
    const triangle = triangleDoc.count;
    const expected = ((linear + 1) * linear) / 2;
    if (triangle !== expected) {
      throw new Error(
        `Invariant violated: linear=${linear}, triangle=${triangle}, expected=${expected}`,
      );
    }
    return { linear, triangle };
  },
});
